src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+    return Types::Viewport::FromDimensions(viewInPixels.Origin(), COORD{ gsl::narrow_cast<short>(viewInPixels.Width() / _api.fontMetrics.cellSize.x), gsl::narrow_cast<short>(viewInPixels.Height() / _api.fontMetrics.cellSize.y) });
+}
+
+[[nodiscard]] Microsoft::Console::Types::Viewport AtlasEngine::GetViewportInPixels(const Types::Viewport& viewInCharacters) const noexcept
+{
+    assert(_api.fontMetrics.cellSize.x != 0);
+    assert(_api.fontMetrics.cellSize.y != 0);
+    return Types::Viewport::FromDimensions(viewInCharacters.Origin(), COORD{ gsl::narrow_cast<short>(viewInCharacters.Width() * _api.fontMetrics.cellSize.x), gsl::narrow_cast<short>(viewInCharacters.Height() * _api.fontMetrics.cellSize.y) });
+}
+
+void AtlasEngine::SetAntialiasingMode(const D2D1_TEXT_ANTIALIAS_MODE antialiasingMode) noexcept
+{
+    const auto mode = gsl::narrow_cast<u16>(antialiasingMode);
+    if (_api.antialiasingMode != mode)
+    {
+        _api.antialiasingMode = mode;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Font);
+    }
+}
+
+void AtlasEngine::SetCallback(std::function<void()> pfn) noexcept
+{
+    _api.swapChainChangedCallback = std::move(pfn);
+}
+
+void AtlasEngine::SetDefaultTextBackgroundOpacity(const float opacity) noexcept
+{
+    const auto mixin = opacity == 1.0f ? 0xff000000 : 0x00000000;
+    if (_api.backgroundOpaqueMixin != mixin)
+    {
+        _api.backgroundOpaqueMixin = mixin;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::SwapChain);
+    }
+}
+
+void AtlasEngine::SetForceFullRepaintRendering(bool enable) noexcept
+{
+}
+
+[[nodiscard]] HRESULT AtlasEngine::SetHwnd(const HWND hwnd) noexcept
+{
+    if (_api.hwnd != hwnd)
+    {
+        _api.hwnd = hwnd;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::SwapChain);
+    }
+    return S_OK;
+}
+
+void AtlasEngine::SetPixelShaderPath(std::wstring_view value) noexcept
+{
+}
+
+void AtlasEngine::SetRetroTerminalEffect(bool enable) noexcept
+{
+}
+
+void AtlasEngine::SetSelectionBackground(const COLORREF color, const float alpha) noexcept
+{
+    const u32 selectionColor = (color & 0xffffff) | gsl::narrow_cast<u32>(std::lroundf(alpha * 255.0f)) << 24;
+    if (_api.selectionColor != selectionColor)
+    {
+        _api.selectionColor = selectionColor;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Settings);
+    }
+}
+
+void AtlasEngine::SetSoftwareRendering(bool enable) noexcept
+{
+}
+
+void AtlasEngine::SetIntenseIsBold(bool enable) noexcept
+{
+}
+
+void AtlasEngine::SetWarningCallback(std::function<void(HRESULT)> pfn) noexcept
+{
+    _api.warningCallback = std::move(pfn);
+}
+
+[[nodiscard]] HRESULT AtlasEngine::SetWindowSize(const SIZE pixels) noexcept
+{
+    u16x2 newSize;
+    RETURN_IF_FAILED(vec2_narrow(pixels.cx, pixels.cy, newSize));
+
+    // At the time of writing:
+    // When Win+D is pressed a render pass is initiated. As conhost is in the background, GetClientRect will return {0,0}.

```

Comment:
>// When Win+D is pressed a r [](http://example.com/codeflow?start=4&length=28)

Only win+D? not also minimize or something like that? win+down arrow?

Reply:
I didn't check that. I only noticed this crash by coincidence one day. Should I check?

Reply:
down arrow I think at least, please.

Reply:
Nope it specifically happens if you press Win+D.
If I press Win+D, `TriggerRedrawCursor` is called exactly once after the window was minimized.
When that happens and a render pass runs, I call `GetClientRect` and call `SetWindowSize` with the result. Since the window is minimized the rect is 0,0.